src/relay/ir/dataflow_matcher.cc

```diff
@@ -0,0 +1,382 @@
+            if (is_pattern_op(arg_node, "divide")) {
+              if (is_expr_op(expr, "divide")) {
+                if (is_expr_op(call_node->args[0], "multiply") ||
+                    is_expr_op(call_node->args[1], "multiply")) {
+                  auto mul =
+                      CallPatternNode::make(op->op, {arg_node->args[0], op->args[(arg_id + 1) % 2]},
+                                            op->attrs, op->type_args);
+                  auto div = CallPatternNode::make(arg_node->op, {mul, arg_node->args[1]},
+                                                   arg_node->attrs, arg_node->type_args);
+                  return VisitDFPattern(div, expr);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return false;
+}
+bool DFPatternMatcher::VisitDFPattern_(const DominatorPatternNode* op, const Expr& expr) {
+  auto watermark = matched_nodes_.size();
+  if (VisitDFPattern(op->child, expr)) {
+    bool matches = true;
+    std::unordered_set<Expr, ObjectHash, ObjectEqual> dominated_exprs;
+    auto child_graph = CreateIndexedGraph(op->child);
+    for (auto node : child_graph.topological_order_) {
+      if (node->ref_.as<WildcardPatternNode>()) {
+        continue;
+      }
+      if (node->dominator_parent_ && node->dominator_parent_->ref_ == op->child) {
+        dominated_exprs.insert(memo_[node->ref_]);
+      }
+    }
+    ClearMap(watermark);
+    auto expr_graph = CreateIndexedGraph(expr);
+    for (auto node : expr_graph.topological_order_) {
+      if (node->dominator_parent_ && node->dominator_parent_->ref_ == expr) {
+        if (dominated_exprs.count(node->ref_) == 0) {
+          bool node_matches = VisitDFPattern(op->parent, node->ref_);
+          ClearMap(watermark);
+          matches = node_matches || VisitDFPattern(op->path, node->ref_);
+          ClearMap(watermark);
+          if (!matches) {
+            return false;
+          }
+        }
+      }
+    }
+    return matches;
+  }
+  return false;
+}
+bool DFPatternMatcher::VisitDFPattern_(const ExprPatternNode* op, const Expr& expr) {
+  return StructuralEqual()(op->expr, expr);
+}
+bool DFPatternMatcher::VisitDFPattern_(const TupleGetItemPatternNode* op, const Expr& expr) {
+  bool matches = false;
+  if (const auto* tuple_get_item_node = expr.as<TupleGetItemNode>()) {
+    matches = (op->index == tuple_get_item_node->index) &&
+              VisitDFPattern(op->tuple, tuple_get_item_node->tuple);
+  }
+  return matches;
+}
+bool DFPatternMatcher::VisitDFPattern_(const TuplePatternNode* op, const Expr& expr) {
+  bool matches = false;
+  if (const auto* tuple_node = expr.as<TupleNode>()) {
+    if (op->fields.size() == tuple_node->fields.size()) {
+      matches = true;
+      size_t i = 0;
+      while (matches && i < op->fields.size()) {
+        matches &= VisitDFPattern(op->fields[i], tuple_node->fields[i]);
+        ++i;
+      }
+    }
+  }
+  return matches;
+}
+Expr InferType(const Expr& expr) {
```

Comment:
This seems useful outside of pattern matching. In python frontend we also use this API a lot. Maybe move it to common utilities?

Reply:
Good idea. Do you have a suggestion for a place? Maybe analysis.h and analysis/util.cc?

Reply:
Sounds good to me. But there is already InferType function with different signature declared in transform.h 
https://github.com/apache/incubator-tvm/blob/master/include/tvm/relay/transform.h#L385

We can also put the new decl there and define it in type_infer.cc. We can reuse `InferType(const Expr& expr, const IRModule& mod)` there.

Reply:
I used to have a Python similar helper in C++ to run passes on an expression and return the updated expression. I believe ppl didn't like it because it was against the module-in and module-out style transformations. I think the one in Python side is mainly used by tests.